controllers/topology/reconcile_state.go

```diff
@@ -16,10 +16,187 @@ limitations under the License.
+}
+
+// reconcileReferencedTemplate reconciles the desired state of the referenced template.
+// NOTE: According to Cluster API operational practices, when a template changes it is required
+// to create a new one, update the reference, and delete the old template (template rotation);
+// this function specifically takes care of the first step and setups everything for executing
+// the other steps.
+func (r *ClusterReconciler) reconcileReferencedTemplate(ctx context.Context, in reconcileReferencedTemplateInput) (func() error, error) {
+	log := ctrl.LoggerFrom(ctx)
+	cleanupFunc := func() error { return nil }
+
+	// If there is no current object, create it.
+	if in.current == nil {
+		log.Info("creating", in.desired.GroupVersionKind().String(), in.desired.GetName())
+		if err := r.Client.Create(ctx, in.desired.DeepCopy()); err != nil {
+			return nil, errors.Wrapf(err, "failed to create the %s object with name %s", in.desired.GetKind(), in.desired.GetName())
+		}
+		return cleanupFunc, nil
+	}
+
+	// Check if the current and desired referenced object are compatible.
+	if err := in.compatibilityChecker(in.current, in.desired); err != nil {
+		return cleanupFunc, err
+	}
+
+	// Check differences between current and desired state, and eventually start template rotation.
+	patchHelper, err := newMergePatchHelper(in.current, in.desired, r.Client)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create patch helper for the %s object with name %s", in.current.GetKind(), in.current.GetName())
+	}
+
+	if patchHelper.HasChanges() {
+		// Create the new template.
+		// NOTE: it is required to assign a new name, because compute the desired object name is enforced to be equal to the current one.
+		oldName := in.current.GetName()
+		newName := in.templateNamer()
+		log.Info(fmt.Sprintf("rotating %s", in.desired.GroupVersionKind().String()), "OldName", oldName, "NewName", newName)
+
+		in.desired.SetName(newName)
+		if err := r.Client.Create(ctx, in.desired); err != nil {
+			return nil, errors.Wrapf(err, "failed to create the %s object with name %s", in.desired.GetKind(), in.desired.GetName())
+		}
+
+		// Update the reference with the new name.
+		// NOTE: Updating the object hosting reference to the template is executed outside this func.
+		in.currentRef.Name = newName
+
+		// Setup a cleanup func for removing the old template.
+		// NOTE: This function must be called after updating the object hosting reference to the template.
+		cleanupFunc = func() error {
+			if err := r.Client.Delete(ctx, in.current); err != nil {
+				return errors.Wrapf(err, "failed to delete the %s object with name %s", in.current.GetKind(), in.current.GetName())
+			}
+			return nil
+		}
+	}
+	return cleanupFunc, nil
+}
+
+// checkReferencedObjectsAreCompatible check if two referenced objects are compatible, meaning that
+// they are of the same GroupKind.
+func checkReferencedObjectsAreCompatible(current, desired client.Object) error {
```

Comment:
Fine for me as is, just a suggestion. What do you think about collecting all errors (I think we use aggregate for this sometimes) and then returning them together? (also in checkReferencedObjectsAreStrictlyCompatible, but in this case across both funcs)

I think this would be good as the user would then get all those compatibility errors at once.

Reply:
We can eventually iterate this when doing #5067; 
Also, even if a safeguard in reconcile will always be required, it is still not fully clear how much of those errors we can prevent in advance, e.g with web hooks.

Reply:
Sounds fine to me